src/charm.py

```diff
@@ -1,19 +1,479 @@
+
+        if not self.is_integrated_to_config_server():
+            logger.info(
+                "Missing integration to config-server. mongos cannot run unless connected to config-server."
+            )
+            self.status.set_and_share_status(
+                BlockedStatus("Missing relation to config-server.")
+            )
+            return
+
+        self.status.set_and_share_status(ActiveStatus())
+
+    # END: hook functions
+
+    # BEGIN: helper functions
+    def is_integrated_to_config_server(self) -> bool:
+        """Returns True if the mongos application is integrated to a config-server."""
+        return self.model.relations[Config.Relations.CLUSTER_RELATIONS_NAME] is not None
+
+    def _get_mongos_config_for_user(
+        self, user: MongoDBUser, hosts: Set[str]
+    ) -> MongosConfiguration:
+        return MongosConfiguration(
+            database=user.get_database_name(),
+            username=user.get_username(),
+            password=self.get_secret(APP_SCOPE, user.get_password_key_name()),
+            hosts=hosts,
+            port=Config.MONGOS_PORT,
+            roles=user.get_roles(),
+            tls_external=None,  # Future PR will support TLS
+            tls_internal=None,  # Future PR will support TLS
+        )
+
+    def get_secret(self, scope: str, key: str) -> Optional[str]:
+        """Get secret from the secret storage."""
+        label = generate_secret_label(self, scope)
+        secret = self.secrets.get(label)
+        if not secret:
+            return
+
+        value = secret.get_content().get(key)
+        if value != Config.Secrets.SECRET_DELETED_LABEL:
+            return value
+
+    def set_secret(self, scope: str, key: str, value: Optional[str]) -> Optional[str]:
+        """Set secret in the secret storage.
+
+        Juju versions > 3.0 use `juju secrets`, this function first checks
+          which secret store is being used before setting the secret.
+        """
+        if not value:
+            return self.remove_secret(scope, key)
+
+        label = generate_secret_label(self, scope)
+        secret = self.secrets.get(label)
+        if not secret:
+            self.secrets.add(label, {key: value}, scope)
+        else:
+            content = secret.get_content()
+            content.update({key: value})
+            secret.set_content(content)
+        return label
+
+    def remove_secret(self, scope, key) -> None:
+        """Removing a secret."""
+        label = generate_secret_label(self, scope)
+        secret = self.secrets.get(label)
+
+        if not secret:
+            return
+
+        content = secret.get_content()
+
+        if not content.get(key) or content[key] == Config.Secrets.SECRET_DELETED_LABEL:
+            logger.error(
+                f"Non-existing secret {scope}:{key} was attempted to be removed."
+            )
+            return
+
+        content[key] = Config.Secrets.SECRET_DELETED_LABEL
+        secret.set_content(content)
```

Comment:
Yes - we should. 

[Some context is that we are currently behind in the world of secrets and we must get all four charms up to date.](https://github.com/canonical/mongodb-operator/pull/333#pullrequestreview-2177627385) This would be a good opportunity to unite this work into shared code. Lets keep this in mind when it comes time to work on that task.